Automating ancil file download

taylorbell57 opened the pull request:
Closes #143 by using code already written in the jwst pipeline to determine the needed reference files and download them

kevin218 commented on eureka/S3_data_reduction/bright2flux.py:
```diff
@@ -183,19 +184,30 @@ def convert_to_e(data, meta, log):
         # Convert from brightness units (MJy/sr) to DNs
-        log.writelog('  Converting from brightness units (MJy/sr) to electrons')
-        meta.photfile = os.path.join(meta.topdir, *meta.ancildir.split(os.sep)) + '/' + data.mhdr['R_PHOTOM'][7:]
+        log.writelog('  Converting from brightness units (MJy/sr) to electrons per second (e/s)')
```
This function is intended to return electrons, not electrons/sec.

taylorbell57 replied:
Huh, well HST data comes in with e/s (or so the FITS header claims) and the dn2electrons function does not multiply by the integration time, so there might be more issues here than just the log messages

kevin218 replied:
Look at lines 65-67 `hst_scan.py`.  When the files are read in, the flux is multiplied by the sample time.

kevin218 commented on eureka/S3_data_reduction/bright2flux.py:
```diff
@@ -183,19 +184,30 @@ def convert_to_e(data, meta, log):
     meta:   MetaClass
         The metadata object.
     """
+    if data.shdr['BUNIT'] != 'ELECTRONS/S':
```
If the units are already in electrons/s, why do we need to download any files?

taylorbell57 replied:
The statement is !=, so no files are downloaded if the units are already good.

kevin218 replied:
so it is!

kevin218 commented on eureka/S3_data_reduction/bright2flux.py:
```diff
@@ -183,19 +184,30 @@ def convert_to_e(data, meta, log):
+    
     return data, meta
+
+def download_ancil(fitsname):
```
We should not require internet access for users to be able to reduce data, so these downloaded files need to go somewhere (like an ancil directory) that can be checked later.

taylorbell57 replied:
The code uses the CRDS variables set in your environment (.bashrc file or equivalent), and it always checks to see if you have the required files first before trying to download anything from the internet. If you have the required files, no internet connection required. The files get saved wherever your CRDS cache is setup to be, which saves you the hassle of redownloading files if you ran stage 2 which would've downloaded the files already. The slight inconvenience is that it checks that you have the right gain and photom files for every file you reduce, but that could also be a feature if you have multiple visits or instruments in the same folder.

kevin218 replied:
Got it.  There's a lot going on under the hood here that isn't obvious.  Consider adding more comments and renaming `download_ancil` to something like `retrieve_ancil` so that there's no confusion.

kevin218 reviewed (changes requested):
Unless I've missed something, it looks like Eureka! is downloading the gain and photom files every time it runs.  The most robust algorithm would be to check for files in the ancil directory.  If a file is missing, download and save it to that directory.  This way, an internet connection isn't required every time someone wants to reduce some data.  If the downloaded files are automatically saved in `crds_cache` then we need to be checking there first.

kevin218 commented:
Before we completely ditch the `ancil` directory, is there any other reason someone would want to use it in the future?  For example, I put some forward models in there for S6.  Maybe someone will want to put stellar Phoenix models in there.

taylorbell57 commented:
> Before we completely ditch the `ancil` directory, is there any other reason someone would want to use it in the future? For example, I put some forward models in there for S6. Maybe someone will want to put stellar Phoenix models in there.

Fair enough, but I don't see how those files would be relevant to S3, so why still specify the ancildir in that stage in particular? And the S6 code doesn't need to know an ancildir - it just looks in the path specified by topdir+model_spectrum (as set in S6). There's nothing saying you can't make an ancilliary files folder, but as of right now there's no code that needs it

kevin218 commented:
> Fair enough, but I don't see how those files would be relevant to S3, so why still specify the ancildir in that stage in particular? And the S6 code doesn't need to know an ancildir - it just looks in the path specified by topdir+model_spectrum (as set in S6). There's nothing saying you can't make an ancilliary files folder, but as of right now there's no code that needs it

You're right, those are later on.  Ok, let's move forward without ancil and see what happens.

taylorbell57 commented:
Alright, I partially resolve some issues and have addressed all your comments above. I've hopefully addressed Issue #121 and at least added error messages to alert users to Issue #169.

taylorbell57 commented:
I just added a few small tweaks to also close https://github.com/kevin218/Eureka/issues/154 and maybe also https://github.com/kevin218/Eureka/issues/155, but I don't really know what's causing https://github.com/kevin218/Eureka/issues/155 so I can't guarantee that issue will be closed (we'll just have to wait and see if anyone encounters that issue again)
